pkg/controller/deployment/deployment_controller.go

```diff
@@ -1294,3 +1413,27 @@ func (dc *DeploymentController) rollbackToTemplate(deployment *extensions.Deploy
 	d, err = dc.updateDeploymentAndClearRollbackTo(deployment)
 	return
 }
+
+// isScalingEvent checks whether the provided deployment has been updated with a scaling event
+// by looking at the desired-replicas annotation in the active replica sets of the deployment.
+func (dc *DeploymentController) isScalingEvent(d *extensions.Deployment) bool {
+	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(d, false)
+	if err != nil {
+		return false
+	}
+	allRSs := append(oldRSs, newRS)
+	for _, rs := range controller.FilterActiveReplicaSets(allRSs) {
+		if desiredString, ok := rs.Annotations[deploymentutil.DesiredReplicasAnnotation]; ok {
```

Comment:
> The rollout path will scale up only the new replica set, proportional scaling distributes for mitigating risk. This was asked actually by @bgrant0607

Yes I agree this is the correct thing todo

> Also, in the future we will need to report progress for a rollout (progressDeadlineSeconds will keep track of intervals between scaling events happening in a rollout and fail a deployment that has no progress). Having external scaling events separate will help in that regard.

I will need to learn more about this to understand.

> I think clients of a watch will receive events on sequence, otherwise watches are broken, which is a bigger issue.

We are not doing anything with the watch event other than using it is a signal that the deployment has needs to be checked that it is fully in sync.  We don't use the context of what was changed in the watch to make any decisions.  Instead we read the latest value from etcd.  This is the correct thing to do, but it means we can not count on just seeing a single event.  We need to look at the whole deployment + related objects state and figure out what works needs to be done.